File: q0005/q0005_dp_iterative.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) <= 1: return s
        maxLen = 1
        l, r = 0, 0

        # dp(j, i) represents if substring(j, i) is a valid palindromic substring
        dp = [[0 for _ in range(len(s))] for _ in range(len(s))]

        for i in range(len(s)):
            # one character itself is a substring
            dp[i][i] = 1

            # loop through all indices before i, and look for palindromes
            for j in range(i):

                # if s[i] == s[j], and substring(j+1, i-1) is a substring as well
                # we found a new palindrome, and hence should potentiall update maxLen
                if (s[i] == s[j] and ((i - j <= 2 or dp[j+1][i-1] == 1))):
                    dp[j][i] = 1
                    currLen = i - j + 1
                    if currLen > maxLen: 
                        maxLen = currLen
                        l = j
                        r = i
        return s[l : r+1]

File: q0005/q0005_dp_iterative.java
class Solution {
    public String longestPalindrome(String s) {
        if (s.length() <= 1) {
            return s;
        }

        int maxLen = 1;
        int l = 0;
        int r = 0;

        // dp(j, i) represents if substring(j, i) is a valid palindromic substring
        int[][] dp = new int[s.length()][s.length()];

        for (int i = 0; i < s.length(); i++) {
            // one character itself is a substring
            dp[i][i] = 1;

            // loop through all indices before i, and look for palindromes
            for (int j = 0; j < i; j++) {
                // if s[i] == s[j], and substring(j+1, i-1) is a substring as well
                // we found a new palindrome, and hence should potentiall update maxLen
                if (s.charAt(i) == s.charAt(j) && ((i - j <= 2 || dp[j + 1][i - 1] == 1))) {
                    dp[j][i] = 1;
                    int currLen = i - j + 1;
                    if (currLen > maxLen) {
                        maxLen = currLen;
                        l = j;
                        r = i;
                    }
                }
            }
        }
        return s.substring(l, r + 1);
    }
}